Content.Server/Pinpointer/ServerPinpointerSystem.cs

```diff
@@ -0,0 +1,157 @@
+        }
+
+        private void OnUseInHand(EntityUid uid, PinpointerComponent component, UseInHandEvent args)
+        {
+            TogglePinpointer(uid, component);
+
+            // try to find target from whitelist
+            if (component.IsActive && component.Whitelist != null)
+            {
+                var target = FindTargetFromWhitelist(uid, component.Whitelist);
+                SetTarget(uid, target, component);
+            }
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            // because target or pinpointer can move
+            // we need to update pinpointers arrow each frame
+            foreach (var uid in ActivePinpointers)
+            {
+                UpdateDirectionToTarget(uid);
+            }    
+        }
+
+        /// <summary>
+        ///     Try to find the closest entity from whitelist on a current map
+        ///     Will return null if can't find anything
+        /// </summary>
+        private EntityUid? FindTargetFromWhitelist(EntityUid uid, EntityWhitelist whitelist,
+            ITransformComponent? transform = null)
+        {
+            if (!Resolve(uid, ref transform))
+                return null;
+
+            var mapId = transform.MapID;
+            var ents = _entityLookup.GetEntitiesInMap(mapId);
+
+            // sort all entities in distance increasing order
+            var l = new SortedList<float, EntityUid>();
+            foreach (var e in ents)
+            {
+                if (whitelist.IsValid(e))
+                {
+                    var dist = (e.Transform.WorldPosition - transform.WorldPosition).LengthSquared;
+                    l.TryAdd(dist, e.Uid);
+                }
+            }
+
+            // return uid with a smallest distacne
+            return l.Count > 0 ? l.First().Value : null;
+        }
+
+        /// <summary>
+        ///     Set pinpointers target to track
+        /// </summary>
+        public void SetTarget(EntityUid uid, EntityUid? target, PinpointerComponent? pinpointer = null)
+        {
+            if (!Resolve(uid, ref pinpointer))
+                return;
+
+            if (pinpointer.Target == target)
+                return;
+
+            pinpointer.Target = target;
+            if (pinpointer.IsActive)
+                UpdateDirectionToTarget(uid, pinpointer);
+        }
+
+        /// <summary>
+        ///     Update direction from pinpointer to selected target (if it was set)
+        /// </summary>
+        private void UpdateDirectionToTarget(EntityUid uid, PinpointerComponent? pinpointer = null)
+        {
+            if (!Resolve(uid, ref pinpointer))
+                return;
+
+            var target = pinpointer.Target;
+            if (target == null || !EntityManager.EntityExists(target.Value))
+            {
+                SetDirection(uid, Direction.Invalid, pinpointer);
```

Comment:
should probably also `SetDistance(uid, Distance.UNKNOWN, pinpointer);` here, right?

Reply:
Yep, fixed it